### PR TITLE
proto 0.12: Require zero bytes for zero query arguments

### PR DESCRIPTION
**Pull request by fmoor**

### CODE REVIEW

**Comment by fmoor** [comment]
Related changes have been made in edgedb-python but need to be released.

File: edb/server/protocol/binary.pyx [hunk 1]
```diff
@@ -2050,8 +2050,15 @@ cdef class EdgeConnection:
...
-            recv_args = 0
+        if self.protocol_version >= (0, 12):
+            if frb_get_len(&in_buf) == 0:
+                recv_args = 0
+            else:
+                recv_args = hton.unpack_int32(frb_read(&in_buf, 4))
+                if recv_args == 0:
+                    raise errors.ProtocolError(
+                        "zero arguments must be encoded as the zero type "
+                        "00000000-0000-0000-0000-000000000000 i.e. zero bytes")
```
Comment (1st1) [inline edb/server/protocol/binary.pyx:2061]: May I suggest another wording:

```suggestion
                        "absence of query arguments must be encoded with a 'zero' type "
                        "(id: 00000000-0000-0000-0000-000000000000, "
                        "encoded with zero bytes)")
```

File: edb/server/protocol/binary.pyx [hunk 2]
```diff
@@ -2050,8 +2050,15 @@ cdef class EdgeConnection:
         # for empty tuple it's okay to send zero-length arguments
         # TODO(tailhook) force zero-length arguments for protocol 0.12 after
         # upgrading the bindings
-        if frb_get_len(&in_buf) == 0:
-            recv_args = 0
+        if self.protocol_version >= (0, 12):
+            if frb_get_len(&in_buf) == 0:
```
Comment (1st1) [inline edb/server/protocol/binary.pyx:2054]: I think `CompiledQuery` must have the input type descriptor -- we need to also check that it's zero type.

File: edb/server/protocol/binary.pyx [hunk 3]
```diff
@@ -2050,8 +2050,15 @@ cdef class EdgeConnection:
         # for empty tuple it's okay to send zero-length arguments
         # TODO(tailhook) force zero-length arguments for protocol 0.12 after
```
Comment (tailhook) [inline edb/server/protocol/binary.pyx:2051]: This TODO should be removed now :)

**Comment by tailhook** [comment]
Also upgrade of python bindings (otherwise tests expectedly fail):
```
diff --git a/setup.py b/setup.py
index 574f92286..87a5d888f 100644
--- a/setup.py
+++ b/setup.py
@@ -41,7 +41,7 @@ except ImportError:
 
 
 RUNTIME_DEPS = [
-    'edgedb>=0.18.0a1',
+    'edgedb>=0.18.0a2',
 
     'asyncpg~=0.24.0',
     'httptools>=0.3.0',
```
But release should be done first:
https://github.com/edgedb/edgedb-python/pull/240

File: edb/server/protocol/binary.pyx (hunk 2 above)
Comment (fmoor) [reply edb/server/protocol/binary.pyx]: changed

File: edb/server/protocol/binary.pyx (hunk 3 above)
Comment (fmoor) [reply edb/server/protocol/binary.pyx]: changed

File: edb/server/protocol/binary.pyx [hunk 4]
```diff
@@ -2048,10 +2049,19 @@ cdef class EdgeConnection:
...
+            frb_len = frb_get_len(&in_buf)
+            is_null_type = \
+                query.query_unit.in_type_id == sertypes.NULL_TYPE_ID.bytes
+            if frb_len and is_null_type:
+                raise errors.ProtocolError(
+                    "absence of query arguments must be encoded with a 'zero' "
+                    "type (id: 00000000-0000-0000-0000-000000000000, "
+                    "encoded with zero bytes)")
+            elif frb_len == 0:
+                recv_args = 0
```
Comment (1st1) [inline edb/server/protocol/binary.pyx:2062]: Hm, you should check if it's `not is_null_type` here and throw an error. You should have 4 states total for `frb_len == 0`, `frb_len > 0`, `is_null_type == true` and `is_null_type == false` and only `is_null_type && not frb_len` is valid out of those.

Comment (fmoor) [reply edb/server/protocol/binary.pyx:2066]: changed

Comment (1st1) [reply edb/server/protocol/binary.pyx:2066]: Please add `assert is_null_type` here for clarity.

File: edb/server/protocol/binary.pyx [hunk 5]
```diff
@@ -2048,10 +2049,23 @@ cdef class EdgeConnection:
...
+                    "type (id: 00000000-0000-0000-0000-000000000000, "
+                    "encoded with zero bytes)")
+            elif not frb_len and not is_null_type:
+                raise errors.ProtocolError(
+                    f"insufficient data for type-id "
+                    f"{query.query_unit.in_type_id}")
+            elif frb_len == 0:
+                recv_args = 0
+            else:
+                recv_args = hton.unpack_int32(frb_read(&in_buf, 4))
```
Comment (1st1) [inline edb/server/protocol/binary.pyx:2068]: Please add a check that the input type id is a shape.

Comment (1st1) [reply edb/server/protocol/binary.pyx:2068]: (I just want this code path to be as easy to follow as possible, sorry for nit-picking)

Comment (fmoor) [reply edb/server/protocol/binary.pyx:2068]: Is there a better way to do this than parsing `query.query_unit.in_type_data`?

File: edb/server/protocol/binary.pyx (hunk 4 above)
Comment (fmoor) [reply edb/server/protocol/binary.pyx:2060]: changed

File: edb/server/protocol/binary.pyx (hunk 5 above)
Comment (1st1) [reply edb/server/protocol/binary.pyx:2068]: Oh, right, it's fine. Please fix the CI.

**Review by 1st1 — APPROVED** [review]

File: edb/server/protocol/binary.pyx (hunk 5 above)
Comment (fmoor) [reply edb/server/protocol/binary.pyx:2068]: > Please fix the CI.

Please approve this PR https://github.com/edgedb/edgedb-python/pull/240.

**Comment by elprans** [comment]
Can this be merged now?

**Pull request merged**